shell/platform/darwin/macos/framework/Source/FlutterSurfaceManager.mm

```diff
@@ -66,48 +66,63 @@ - (void)ensureSurfaceSize:(CGSize)size {
       std::make_unique<MacOSSwitchableGLContext>(openGLContext));
 
   for (int i = 0; i < kBufferCount; ++i) {
-    if (_ioSurface[i]) {
-      CFRelease(_ioSurface[i]);
-    }
-    unsigned pixelFormat = 'BGRA';
-    unsigned bytesPerElement = 4;
-
-    size_t bytesPerRow =
-        IOSurfaceAlignProperty(kIOSurfaceBytesPerRow, size.width * bytesPerElement);
-    size_t totalBytes = IOSurfaceAlignProperty(kIOSurfaceAllocSize, size.height * bytesPerRow);
-    NSDictionary* options = @{
-      (id)kIOSurfaceWidth : @(size.width),
-      (id)kIOSurfaceHeight : @(size.height),
-      (id)kIOSurfacePixelFormat : @(pixelFormat),
-      (id)kIOSurfaceBytesPerElement : @(bytesPerElement),
-      (id)kIOSurfaceBytesPerRow : @(bytesPerRow),
-      (id)kIOSurfaceAllocSize : @(totalBytes),
-    };
-    _ioSurface[i] = IOSurfaceCreate((CFDictionaryRef)options);
-
-    glBindTexture(GL_TEXTURE_RECTANGLE_ARB, _backingTexture[i]);
-
-    CGLTexImageIOSurface2D(CGLGetCurrentContext(), GL_TEXTURE_RECTANGLE_ARB, GL_RGBA,
-                           int(size.width), int(size.height), GL_BGRA, GL_UNSIGNED_INT_8_8_8_8_REV,
-                           _ioSurface[i], 0 /* plane */);
-    glBindTexture(GL_TEXTURE_RECTANGLE_ARB, 0);
-
-    glBindFramebuffer(GL_FRAMEBUFFER, _frameBufferId[i]);
-    glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_RECTANGLE_ARB,
-                           _backingTexture[i], 0);
-
-    FML_DCHECK(glCheckFramebufferStatus(GL_FRAMEBUFFER) == GL_FRAMEBUFFER_COMPLETE);
+    [self backTextureWithIOSurface:&_ioSurface[i]
+                              size:size
+                    backingTexture:_backingTexture[i]
+                               fbo:_frameBufferId[i]];
   }
 }
 
-- (void)swapBuffers {
+- (void)backTextureWithIOSurface:(IOSurfaceRef*)ioSurface
+                            size:(CGSize)size
+                  backingTexture:(GLuint)texture
+                             fbo:(GLuint)fbo {
+  if (ioSurface != nullptr && *ioSurface) {
+    CFRelease(*ioSurface);
+  }
+  unsigned pixelFormat = 'BGRA';
+  unsigned bytesPerElement = 4;
+
+  size_t bytesPerRow = IOSurfaceAlignProperty(kIOSurfaceBytesPerRow, size.width * bytesPerElement);
+  size_t totalBytes = IOSurfaceAlignProperty(kIOSurfaceAllocSize, size.height * bytesPerRow);
+  NSDictionary* options = @{
+    (id)kIOSurfaceWidth : @(size.width),
+    (id)kIOSurfaceHeight : @(size.height),
+    (id)kIOSurfacePixelFormat : @(pixelFormat),
+    (id)kIOSurfaceBytesPerElement : @(bytesPerElement),
+    (id)kIOSurfaceBytesPerRow : @(bytesPerRow),
+    (id)kIOSurfaceAllocSize : @(totalBytes),
+  };
+  *ioSurface = IOSurfaceCreate((CFDictionaryRef)options);
```

Comment:
Now looking at the method name and the implementation, I think hiding the `iosSurface` pointer reassignment in the implementation could be error prone. 

Maybe we can refactor this part of the code to something like:
`-(IOSurfaceRef*)createNewIOSurface:(CGSize)size;`

Then the caller becomes:
``` objc
for (int i = 0; i < kBufferCount; ++i) {
  if (_ioSurface[i]) {
    CFRelease(_ioSurface[i]);
  }
  _ioSurface[i] = [self createNewIOSurface:size];
  [self backTextureWithIOSurface:_ioSurface[i] size:size backingTexture:_backingTexture[i] fbo:_frameBufferId[i]];
}
```
WDYT? cc @iskakaushik

Reply:
Yeah that sounds good, I'll do refactor it.

Reply:
Update: I realized because of the use case for backTextureWithIOSurface in FlutterMacOSCompositor 
```  GLuint fbo = [surfaceManager getFramebuffer];
  GLuint texture = [surfaceManager getTexture];
  IOSurfaceRef* io_surface_ref = [surfaceManager getIOSurface];

  CGSize size = CGSizeMake(config->size.width, config->size.height);

  [surfaceManager backTextureWithIOSurface:io_surface_ref size:size backingTexture:texture fbo:fbo];```

The exact refactor you suggested doesn't work but we can refactor it to take an int representing which array index to use, ie _ioSurface[kFlutterSurfaceManagerFrontBuffer]